src/persist-client/src/cache.rs

```diff
@@ -250,10 +262,179 @@ async fn consensus_rtt_latency_task(
     })
 }
 
+trait DynState: Debug + Send + Sync {
+    fn codecs(&self) -> (String, String, String, String, Option<CodecConcreteType>);
+    fn as_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync>;
+}
+
+impl<K, V, T, D> DynState for Mutex<TypedState<K, V, T, D>>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Codec64,
+{
+    fn codecs(&self) -> (String, String, String, String, Option<CodecConcreteType>) {
+        (
+            K::codec_name(),
+            V::codec_name(),
+            T::codec_name(),
+            D::codec_name(),
+            Some(CodecConcreteType(std::any::type_name::<
+                TypedState<K, V, T, D>,
+            >())),
+        )
+    }
+
+    fn as_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync> {
+        self
+    }
+}
+
+/// A cache of `TypedState`, shared between all machines for that shard.
+///
+/// This is shared between all machines that come out of the same
+/// [PersistClientCache], but in production there is one of those per process,
+/// so in practice, we have one copy of state per shard per process.
+///
+/// The mutex contention between commands is not an issue, because if two
+/// command for the same shard are executing concurrently, only one can win
+/// anyway, the other will retry. With the mutex, we even get to avoid the retry
+/// if the racing commands are on the same process.
+#[derive(Debug, Default)]
+pub struct StateCache {
+    states: Mutex<BTreeMap<ShardId, Arc<OnceCell<Weak<dyn DynState>>>>>,
+}
+
+#[derive(Debug)]
+enum StateCacheInit {
+    Init(Arc<dyn DynState>),
+    NeedInit(Arc<OnceCell<Weak<dyn DynState>>>),
+}
+
+impl StateCache {
+    pub(crate) async fn get<K, V, T, D, F, InitFn>(
+        &self,
+        shard_id: ShardId,
+        mut init_fn: InitFn,
+    ) -> Result<Arc<Mutex<TypedState<K, V, T, D>>>, Box<CodecMismatch>>
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+        F: Future<Output = Result<TypedState<K, V, T, D>, Box<CodecMismatch>>>,
+        InitFn: FnMut() -> F,
+    {
+        loop {
+            let init = {
+                let mut states = self.states.lock().await;
```

Comment:
at the risk of overcomplication -- any thoughts on a RwLock here? I'd think the vast majority of accesses will be a read of an `Init` value

Reply:
eh should prolly be benchmarked before making that kind of change